### PR TITLE
[Metrics] Fix/4237 auc unstable reorder

### LINKED ISSUE #4237: [Metrics] AUC reorder is unstable
## 🐛 Bug

When `auc(x, y, reorder=True)` is called (`True` is the default) it reorders the xs and ys. It uses `torch.argsort` on `x` internally, which is unstable and does not take the second key `y` into account when sorting. In some cases this leads to reordering that changes the order of `ys` in an undesirable way resulting in incorrect metric computation.

```
fpr = torch.tensor([0.4167, 0.5000, 0.5000,
        0.5000, 0.5833, 0.6667, 0.6667, 0.6667, 0.7500, 0.7500, 0.7500, 0.8333,
        0.9167, 0.9167, 0.9167, 1.0000])
tpr = torch.tensor([0.1111, 0.1111, 0.2222,
        0.3333, 0.3333, 0.3333, 0.4444, 0.5556, 0.5556, 0.6667, 0.7778, 0.7778,
        0.7778, 0.8889, 1.0000, 1.0000])

auc_sk = torch.tensor(sk_auc(fpr, tpr)).float()
auc_reorder = auc(fpr, tpr, reorder=True)
auc_no_reorder = auc(fpr, tpr, reorder=False)

print("Sklearn", auc_sk.item())
print("PTL no reorder", auc_no_reorder.item(), "SK==PTL:", torch.allclose(auc_sk, auc_no_reorder))
print("PTL with reorder", auc_reorder.item(), "SK==PTL:", torch.allclose(auc_sk, auc_reorder))
print("unstable argsort: ", torch.argsort(fpr))

>>> Sklearn 0.324056476354599
>>> PTL no reorder 0.324056476354599 SK==PTL: True
>>> PTL with reorder 0.3240675926208496 SK==PTL: False
>>> unstable argsort:  tensor([ 0,  1,  2,  3,  4,  7,  6,  5,  8,  9, 10, 11, 12, 13, 14, 15])
```

See an example [colab](https://colab.research.google.com/drive/18bG1cqSf-KJOmvZWKUtGUASZagrDqxf6#scrollTo=S9_kfT5fqBpH). I'm sure there is a more minimal example possible, this is what I encountered when implementing a different issue.

### Expected behavior

Either a stable way of sorting across both `x` and `y` or removing the `reorder` argument entirely.

### Environment

Colab.

### PR DESCRIPTION
**Pull request by ddrevicky**
## What does this PR do?

Deprecates the `reorder` parameter to the `auc` metric because of `torch.argsort` being unstable.

Fixes #4237 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SkafteNicki — COMMENTED** [review]
Please also add note at argument:
https://github.com/PyTorchLightning/pytorch-lightning/blob/d0c96b195f5a22f454c9d73ce595e7911b9dafc1/pytorch_lightning/metrics/functional/classification.py#L808

**Review by tchaton — APPROVED** [review]
Great catch ! I didn't know `argsort` was unstable.

**Review by s-rog — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Comment by SkafteNicki** [comment]
@ddrevicky could you please solve the pep8 issue:
`./pytorch_lightning/metrics/functional/classification.py:841: [F541] f-string is missing placeholders`
